### PR TITLE
Enable custom From-Header for outbound calls

### PR DESCRIPTION
**Pull request by fabrykowski**
Enables setting a From-header different from the Contact-header for an outbound call.

### CODE REVIEW

**Comment by fabrykowski** [comment]
I'm willing to prepare the updated documentation as well, but I don't know where.

**Comment by ibc** [comment]
What is the motivation for this?

**Comment by fabrykowski** [comment]
For example multiple people working in an office with a central reception number. (e.g. support hotline)
All desk phones and VoIP-PSTN-carriers ignore the `display_name`-part of the From-header and display the username instead. So I need a way to send `From: <sip:+1234567890@example.com>;tag=xyz` in the INVITEs, despite REGISTERing as `From: <sip:desk-phone-123@example.com>;tag=abc`.

**Comment by ibc** [comment]
For that, the `P-Preferred-Identity` header must be used as stated by the RFC:

- https://tools.ietf.org/html/rfc3325#section-9.2
- https://tools.ietf.org/html/rfc5876

Sending a random/unknown `From` header is not the way to go.

**Comment by fabrykowski** [comment]
Unfortunately RFC 3325 is only a memo ("This memo provides information for the Internet community.  It does not specify an Internet standard of any kind.") and the PBXs in question only support the official RFC 3261 with the corresponding From-header ([8.1.1.3](https://tools.ietf.org/html/rfc3261#section-8.1.1.3) and [20.20](https://tools.ietf.org/html/rfc3261#section-20.20)).

**Comment by ibc** [comment]
@jmillan thoughts?

**Comment by ibc** [comment]
> Besides: if `anonymous` is set, it does exactly the same.

This is how anonymous call must be done according to the RFCs.

File: lib/RTCSession.js [hunk 1]
```diff
@@ -353,6 +353,17 @@ module.exports = class RTCSession extends EventEmitter
       extraHeaders.push(`P-Preferred-Identity: ${this._ua.configuration.uri.toString()}`);
       extraHeaders.push('Privacy: id');
     }
+    else
+    {
+      if (options.from_uri)
+      {
+        requestParams.from_uri = URI.parse(options.from_uri);
```
Comment (ibc) [inline lib/RTCSession.js:360]: `P-Preferred-Identity` is required if URI in From header does not honor the original one.

```suggestion
        requestParams.from_uri = URI.parse(options.from_uri);
        extraHeaders.push(`P-Preferred-Identity: ${this._ua.configuration.uri.toString()}`);
```

**Review by ibc — CHANGES_REQUESTED** [review]

**Comment by fabrykowski** [comment]
My precise use case:
I am developing a software for front-office employees, who call customers on behalf of other back-office employees. If the customer could not be reached, they should be able to call back the correct back-office employee.
I have run a test with three different PSTN endpoints and all of them ignored the contents of `P-Preferred-Identity`, while displaying the value from the `From` header.

**Comment by jmillan** [comment]
Hi @fabrykowski, I see your scenario and concern. Here some changes to your proposal.

* Use `options.fromUserName` as **string** instead of providing a hole URI. We don't want to change the URI host.
* Slight code cleanup.

```js
    if (anonymous)
    {
      requestParams.from_display_name = 'Anonymous';
      requestParams.from_uri = new URI('sip', 'anonymous', 'anonymous.invalid');

      extraHeaders.push(`P-Preferred-Identity: ${this._ua.configuration.uri.toString()}`);
      extraHeaders.push('Privacy: id');
    }
    else if (options.fromUserName)
    {
      requestParams.from_uri = new URI('sip', options.fromUserName, this._ua.configuration.uri.host);

      extraHeaders.push(`P-Preferred-Identity: ${this._ua.configuration.uri.toString()}`);
    }

    if (options.fromDisplayName)
    {
      requestParams.from_display_name = options.fromDisplayName;
    }

```

**Comment by fabrykowski** [comment]
I like the idea of `options.fromUserName`!

The standalone `if`-block for `fromDisplayName` allows to override the display name even if `anonymous` is set – I like this, because then the name can be localised.

**Review by ibc — APPROVED** [review]
I approve this, but it must first added to the doc website. Can you please do a PR also in there?

**Comment by fabrykowski** [comment]
See versatica/jssip.net#8

**Comment by fabrykowski** [comment]
Thanks!

**Comment by jmillan** [comment]
New version released, thanks @fabrykowski!

**Pull request merged**